### ISSUE TITLE
Row[N] predicates should wrap T argument values in Tools.field() to allow for raw type scalar subqueries

### ISSUE THREAD

**Issue by do4gr** [jOOQ/jOOQ#8220]
### Expected behavior and actual behavior:

We had a jOOQ statement successfully working in 3.11.0 for MySQL and Postgres and it broke for Postgres with 3.11.9.

```
val selectQuery = sql
      .select(orderByField)
      .from(modelTable(model))
      .where(idField.equal(stringDummy))


val cursor = row(orderByFieldWithAlias, idFieldWithAlias).lessThan(selectQuery, stringDummy)

val base = sql
      .select(jooqFields.toVector: _*)
      .from(modelTable(model).as(topLevelAlias))
      .where(condition, cursor)
      .orderBy(order: _*)
```

This was producing the following SQL: 

```sql
select
	  "Alias"."name",
	  "Alias"."id",
	  "Alias"."createdAt"
	from "NonEmbeddedPaginationSpecForCuid"."List" as "Alias"
	where ("Alias"."createdAt", "Alias"."id") > ((
	  select "NonEmbeddedPaginationSpecForCuid"."List"."createdAt"
	  from "NonEmbeddedPaginationSpecForCuid"."List"
	  where "NonEmbeddedPaginationSpecForCuid"."List"."id" = $1
	), $2)
	order by
	  "Alias"."createdAt" asc,
	  "Alias"."id" asc
```

After upgrading to 3.11.9 we get the following runtime error:

```
org.jooq.exception.SQLDialectNotSupportedException: Type class org.jooq.impl.SelectImpl is not supported in dialect POSTGRES
	at org.jooq.impl.DefaultDataType.getDataType(DefaultDataType.java:884)
	at org.jooq.impl.DefaultDataType.getDataType(DefaultDataType.java:823)
	at org.jooq.impl.DefaultBinding$AbstractBinding.sqlCast(DefaultBinding.java:618)
	at org.jooq.impl.DefaultBinding$AbstractBinding.sql(DefaultBinding.java:707)
	at org.jooq.impl.Val.accept(Val.java:80)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:509)
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:446)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:207)
	at org.jooq.impl.RowImpl.accept(RowImpl.java:207)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:509)
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:446)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:207)
	at org.jooq.impl.RowCondition$Native.accept(RowCondition.java:232)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:509)
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:446)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:207)
	at org.jooq.impl.RowCondition.accept(RowCondition.java:106)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:509)
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:446)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:207)
	at org.jooq.impl.ConditionProviderImpl.accept(ConditionProviderImpl.java:125)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:509)
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:446)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:207)
	at org.jooq.impl.ConditionProviderImpl.accept(ConditionProviderImpl.java:125)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:509)
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:446)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:207)
	at org.jooq.impl.SelectQueryImpl.toSQLReference0(SelectQueryImpl.java:1311)
	at org.jooq.impl.SelectQueryImpl.toSQLReference0(SelectQueryImpl.java:1034)
	at org.jooq.impl.SelectQueryImpl.toSQLReferenceLimitDefault(SelectQueryImpl.java:823)
	at org.jooq.impl.SelectQueryImpl.accept(SelectQueryImpl.java:640)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:509)
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:446)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:207)
	at org.jooq.impl.DefaultDSLContext.render(DefaultDSLContext.java:687)
	at org.jooq.impl.AbstractQuery.getSQL(AbstractQuery.java:565)
	at org.jooq.impl.AbstractQuery.getSQL(AbstractQuery.java:555)
	at org.jooq.impl.AbstractDelegatingQuery.getSQL(AbstractDelegatingQuery.java:101)
	at com.prisma.connector.shared.jdbc.SharedSlickExtensions.$anonfun$jooqToDBIO$1(SharedSlickExtensions.scala:57)
	at slick.jdbc.SimpleJdbcAction.run(StreamingInvokerAction.scala:70)
	at slick.jdbc.SimpleJdbcAction.run(StreamingInvokerAction.scala:69)
	at slick.basic.BasicBackend$DatabaseDef$$anon$2.liftedTree1$1(BasicBackend.scala:275)
	at slick.basic.BasicBackend$DatabaseDef$$anon$2.run(BasicBackend.scala:275)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

### Versions:

- jOOQ:  3.11.9
- Scala: 2.12.3
- Database (include vendor): Postgres
- OS: Mac Os
- JDBC Driver (include name if inofficial driver): org.postgresql 42.2.2

**Comment by lukaseder** [comment]
Thank you very much for your report. How did you construct this expression?

```scala
val cursor = row(orderByFieldWithAlias, idFieldWithAlias).lessThan(selectQuery, stringDummy)
```

I'm particularly curious about the inferred or explicit data types of each local variable. There is now method that would accept your arguments to the `lessThan()` method:
https://www.jooq.org/javadoc/latest/org/jooq/Row2.html

But perhaps there's a Scala/Java interop issue?

The workaround is to wrap your `Select` statement in a `Field` using`DSL.field(Select)`, and possibly wrap the `stringDummy` as well...

**Comment by do4gr** [comment]
Here is the file that constructs that expression: 

```scala
trait CursorConditionBuilder extends BuilderBase {
  // This creates a query that checks if the id is in a certain set returned by a subquery Q.
  // The subquery Q fetches all the ID's defined by the cursors and order.
  // On invalid cursor params, no error is thrown. The result set will just be empty.

  def buildCursorCondition(queryArguments: QueryArguments, model: Model): Condition = {
    val (before, after, orderBy) = (queryArguments.before, queryArguments.after, queryArguments.orderBy)
    // If both params are empty, don't generate any query.
    if (before.isEmpty && after.isEmpty) return noCondition()

    val idFieldWithAlias: jooq.Field[AnyRef] = aliasColumn(model.dbNameOfIdField_!)
    val idField: jooq.Field[AnyRef]          = modelIdColumn(model)

    // First, we fetch the ordering for the query. If none is passed, we order by id, ascending.
    // We need that since before/after are dependent on the order.
    val (orderByField: jooq.Field[AnyRef], orderByFieldWithAlias: jooq.Field[AnyRef], sortDirection: String) = orderBy match {
      case Some(order) => (modelColumn(order.field), aliasColumn(order.field.dbName), order.sortOrder.toString)
      case None        => (idField, idFieldWithAlias, "asc")
    }

    val selectQuery = sql
      .select(orderByField)
      .from(modelTable(model))
      .where(idField.equal(stringDummy))

    // Then, we select the comparison operation and construct the cursors. For instance, if we use ascending order, and we want
    // to get the items before, we use the "<" comparator on the column that defines the order.
    def cursorFor(cursorType: String): Condition = (cursorType, sortDirection.toLowerCase.trim) match {
      case ("before", "asc")  => row(orderByFieldWithAlias, idFieldWithAlias).lessThan(field(selectQuery), stringDummy)
      case ("before", "desc") => row(orderByFieldWithAlias, idFieldWithAlias).greaterThan(field(selectQuery), stringDummy)
      case ("after", "asc")   => row(orderByFieldWithAlias, idFieldWithAlias).greaterThan(field(selectQuery), stringDummy)
      case ("after", "desc")  => row(orderByFieldWithAlias, idFieldWithAlias).lessThan(field(selectQuery), stringDummy)
      case _                  => throw new IllegalArgumentException
    }

    val afterCursorFilter: Condition  = after.map(_ => cursorFor("after")).getOrElse(noCondition())
    val beforeCursorFilter: Condition = before.map(_ => cursorFor("before")).getOrElse(noCondition())

    afterCursorFilter.and(beforeCursorFilter)
  }
```

stringDummy is just "" as a placeholder since we at later stage go over the statements and replace all placeholders with the actual parameters.

I tried to wrap the subselect in `field` in the file above but this produced a runtime error as well: 

```
org.jooq.exception.SQLDialectNotSupportedException: Type class org.jooq.impl.ScalarSubquery is not supported in dialect POSTGRES
	at org.jooq.impl.DefaultDataType.getDataType(DefaultDataType.java:884)
	at org.jooq.impl.DefaultDataType.getDataType(DefaultDataType.java:823)
	at org.jooq.impl.DefaultBinding$AbstractBinding.sqlCast(DefaultBinding.java:618)
	at org.jooq.impl.DefaultBinding$AbstractBinding.sql(DefaultBinding.java:707)
	at org.jooq.impl.Val.accept(Val.java:80)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:509)
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:446)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:207)
	at org.jooq.impl.RowImpl.accept(RowImpl.java:207)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:509)
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:446)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:207)
	at org.jooq.impl.RowCondition$Native.accept(RowCondition.java:232)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:509)
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:446)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:207)
	at org.jooq.impl.RowCondition.accept(RowCondition.java:106)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:509)
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:446)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:207)
	at org.jooq.impl.ConditionProviderImpl.accept(ConditionProviderImpl.java:125)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:509)
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:446)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:207)
	at org.jooq.impl.ConditionProviderImpl.accept(ConditionProviderImpl.java:125)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:509)
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:446)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:207)
	at org.jooq.impl.SelectQueryImpl.toSQLReference0(SelectQueryImpl.java:1311)
	at org.jooq.impl.SelectQueryImpl.toSQLReference0(SelectQueryImpl.java:1034)
	at org.jooq.impl.SelectQueryImpl.toSQLReferenceLimitDefault(SelectQueryImpl.java:823)
	at org.jooq.impl.SelectQueryImpl.accept(SelectQueryImpl.java:640)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:509)
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:446)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:207)
	at org.jooq.impl.DefaultDSLContext.render(DefaultDSLContext.java:687)
	at org.jooq.impl.AbstractQuery.getSQL(AbstractQuery.java:565)
	at org.jooq.impl.AbstractQuery.getSQL(AbstractQuery.java:555)
	at org.jooq.impl.AbstractDelegatingQuery.getSQL(AbstractDelegatingQuery.java:101)
	at com.prisma.connector.shared.jdbc.SharedSlickExtensions.$anonfun$jooqToDBIO$1(SharedSlickExtensions.scala:57)
	at slick.jdbc.SimpleJdbcAction.run(StreamingInvokerAction.scala:70)
	at slick.jdbc.SimpleJdbcAction.run(StreamingInvokerAction.scala:69)
	at slick.basic.BasicBackend$DatabaseDef$$anon$2.liftedTree1$1(BasicBackend.scala:275)
	at slick.basic.BasicBackend$DatabaseDef$$anon$2.run(BasicBackend.scala:275)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
``` 

Wrapping the stringDummy as well in `field` fails during our own placeholder replacement step since it is then not recognized as a placeholder and the number of expected args is off.

**Comment by do4gr** [comment]
Your hint got me in the right direction, I got it working again with this combination: 

```
val selectQuery: SelectConditionStep[Record2[AnyRef, AnyRef]] = sql
      .select(orderByField, param())
      .from(modelTable(model))
      .where(idField.equal(param()))

  row(orderByFieldWithAlias, idFieldWithAlias).lessThan(selectQuery)
```

Thanks for the quick reply. Should I still keep the issue open since there was an unexpected change in behavior between 3.11.0 and 3.11.9?

**Comment by lukaseder** [comment]
I still cannot reproduce this particular issue, but a similar issue can be produced using:

```java
@Test
@SuppressWarnings({ "rawtypes", "unchecked" })
public void testRowWithSelect() {

    // [#8220] Using a Select as a bind value with row value expressions
    Field<Integer> s0 = field(selectZero());
    Field<Integer> s1 = field(selectOne());

    assertEquals(
        create.render(row(zero(), one()).eq(s0, s1)),
        create.render(((Row2) row(zero(), one())).eq(selectZero(), selectOne()))
    );
}
```

The test fails with:

```
org.junit.ComparisonFailure: expected:<(0, 1) = ([(select 0 as `zero` from dual), (select 1 as `one` from dual)])> but was:<(0, 1) = ([?, ?])>
	at org.junit.Assert.assertEquals(Assert.java:117)
	at org.junit.Assert.assertEquals(Assert.java:146)
	at org.jooq.test.AbstractTest.assertEquals(AbstractTest.java:201)
	at org.jooq.test.RogueAPIUsageTest.testRowWithSelect(RogueAPIUsageTest.java:136)
```

Clearly, the intent was not to pass two bind variables, but to use the `Select` as scalar subqueries.

**Comment by lukaseder** [comment]
This is a similar issue as #6362, but for `DSL.row()` constructors and their methods.

**Comment by lukaseder** [comment]
I hope this fixes it. Feel free to create a new issue if you think it doesn't.

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.14.
